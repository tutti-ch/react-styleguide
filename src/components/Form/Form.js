import React from "react"
import Input from "./_Input"
import InputCheckbox from "./_InputCheckbox"
import InputRadio from "./_InputRadio"
import Textarea from "./_Textarea"
import Slider from "./_Slider"
import Table from "../Table"
import InputRadioGroup from "./_InputRadioGroup"

/**
 * This component is only used by the style guide.
 * It should not be used in production.
 */
export default () => (
  <Table>
    <Table.Body>
      <Table.Row>
        <Table.Cell>
          <h3>Inputs</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <Input label="Block text input"/>
          <Input label="Inline text input" inline/>
          <Input label="Number" type="number" min="100" max="1000" step="50" value="150"/>
          <Input label="Password" type="password"/>
          <Input label="Email" type="email"/>
          <Input label="Phone" type="phone"/>
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Slider</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <Slider label="Slider"
                  min={1000}
                  max={10000}
                  minValue={1300}
                  maxValue={10000}
                  minRange={500}
                  prefix={["From ", "To "]}
                  suffix=".-"
                  step={100}
                  multiple/>
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Checkbox Group</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <InputCheckbox value="red" label="Red" checked/>
          <InputCheckbox value="blue" label="Blue" checked/>
          <InputCheckbox value="green" label="Green"/>
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Button Group</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <InputRadioGroup>
            <InputRadio checked value={10} label="React.js" name="r1"/>
            <InputRadio value={10} label="Vue.js" name="r1"/>
            <InputRadio value={10} label="Angular.js" name="r1"/>
          </InputRadioGroup>
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Textarea</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <Textarea label="This is a textarea"/>
        </Table.Cell>
      </Table.Row>
    </Table.Body>
  </Table>
)
