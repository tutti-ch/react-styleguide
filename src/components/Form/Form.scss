@import "../styles/main";

@mixin placeholder {
  // IMPORTANT: Do not combine the following placeholder styles, as otherwise they will break.

  // Chrome/Opera/Safari
  ::-webkit-input-placeholder {
    @content
  }

  // Firefox 19+
  :-moz-placeholder {
    @content
  }

  // IE 10+
  ::-moz-placeholder {
    @content
  }

  // Firefox 18-
  :-ms-input-placeholder {
    @content
  }
}

@include placeholder {
  color: $greyscale-60;
}

@mixin focus {
  transition: $transition-m;
  box-shadow: none;

  &:focus {
    outline: none;
    @include mediumShadow();
  }
}

input[type="text"],
input[type="password"],
input[type="email"],
input[type="phone"],
input[type="number"],
textarea {
  border: $border-default;
  background-color: $greyscale-100;
  padding: $size-s;
  border-radius: 2px;
  position: relative;
  box-sizing: border-box;
  width: 100%;
  @include focus();
}

input[type="number"] {
  width: 6rem;
}

textarea {
  min-height: 8rem;
  resize: none;
}

textarea + .label {
  top: $size-s;
  transform: translateY(0);
}

input:focus + .label,
textarea:focus + .label,
.slider .label,
.has-value .label {
  font-size: 0.85rem;
  top: 0;
  left: 0;
  transform: translateY(calc(-100% - 0.25rem));
}

input[type="checkbox"],
input[type="radio"] {
  font: inherit;
  vertical-align: middle;
  background: $greyscale-100;
  border-radius: $size-xxxs / 2;
  box-sizing: border-box;
  border: $border-default;
  position: relative;
  margin-right: $size-m;
}

.wrapper {
  display: block;
  position: relative;
  font-size: 1rem;

  &.inline {
    display: inline-block;
  }

  & + .wrapper {
    margin-top: $size-xl;
  }

  & + label.wrapper {
    margin-top: $size-m;
  }
}

.cb-label {
  user-select: none;
  cursor: pointer;
}

.label {
  position: absolute;
  pointer-events: none;
  left: $size-s;
  top: 50%;
  color: $greyscale-60;
  transform: translateY(-50%);
  transition: $transition-s;
}

.slider {
  position: relative;
  display: inline-block;
  width: 100%;
  height: $size-xxs;
  background: $blue-50;
  border-radius: $size-xxxs / 2;
  border: 0 solid #010101;
  outline: none;

  .label {
    top: -$size-xxxs;
  }

  .labelText {
    display: inline-block;
    margin-right: $size-xs;

    &:after {
      content: ':'
    }
  }

  .rangeMin {
    display: inline-block;
    min-width: 5.5rem;
  }

  .rangeMax {
    margin-left: $size-xs;
  }

  .thumb {
    user-select: none;
    display: inline-block;
    border: $border-default;
    height: $size-ml;
    width: $size-ml;
    border-radius: 50%;
    background: #ffffff;
    outline: none;
    cursor: pointer;
    position: absolute;
    transition: width 0.1s ease-out, height 0.1s ease-out;
    top: 50%;
    margin-left: -$size-ml / 2;
    transform: translateY(-50%);

    &.trans {
      transition: $transition-m;
    }

    &[name=min] {
      margin-left: -1px;
    }

    &[name=max] {
      margin-right: -1px
    }

    &[data-value]:before {
      position: absolute;
      content: attr(data-value);
      left: -50%;
      right: -50%
    }
  }
}