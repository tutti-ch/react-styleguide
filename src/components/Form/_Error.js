import React, { Component } from "react";
import PropTypes from "prop-types";
import classes from "./Form.scss";

export default class Error extends Component {
  static contextTypes = {
    genericErrorHandler: PropTypes.func
  };

  constructor(props, context) {
    super(props);

    this.setError = this.setError.bind(this);
    this.state = {
      error: null
    };

    if (typeof context.genericErrorHandler === "function") {
      context.genericErrorHandler(this);
    }
  }

  setError(message) {
    this.setState({ error: message });
  }

  render() {
    const { error } = this.state;

    if (!error) {
      return null;
    }

    return <div className={classes.genericError}>{error}</div>;
  }
}
