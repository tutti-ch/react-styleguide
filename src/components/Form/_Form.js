import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Input from "./_Input";
import Textarea from "./_Textarea";
import Radio from "./_InputRadio";
import RadioGroup from "./_InputRadioGroup";
import Checkbox from "./_InputCheckbox";
import Select from "./_Select";
import Stepper from "./_Stepper";
// import Slider from "./_Slider"; removed slider as we don't use it at the moment
import Toggle from "./_Toggle";
import Error from "./_Error";
import { filterProps } from "../../helpers/functions";
import classes from "./Form.scss";

export default class Form extends Component {
  static Input = Input;
  static Radio = Radio;
  static RadioGroup = RadioGroup;
  static Checkbox = Checkbox;
  static Select = Select;
  // static Slider = Slider;
  static Textarea = Textarea;
  static GenericError = Error;
  static Toggle = Toggle;
  static Stepper = Stepper;

  static propTypes = {
    className: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.array,
      PropTypes.string
    ]),
    children: PropTypes.node,
    handleSubmit: PropTypes.func // Function to handle submit.
  };

  static childContextTypes = {
    register: PropTypes.func,
    onSubmit: PropTypes.func,
    genericErrorHandler: PropTypes.func
  };

  /**
   * Assign key value pair to the first object. In case a property
   * already exists, this function will aggregate the results.
   *
   * @example
   * arguments: { a: "b" }, { a: "c", d: 1 }
   * returns: { a: ["b", "c"], d: 1 }
   *
   * @param values
   * @param keyValuePair
   */
  static assignKeyPair(values, keyValuePair) {
    Object.keys(keyValuePair).forEach(k => {
      if (values[k]) {
        // Cast to an array ({ a: "b" } becomes { a: ["b"] })
        if (!Array.isArray(values[k])) {
          values[k] = [values[k]];
        }

        values[k].push(keyValuePair[k]);
      } else {
        values[k] = keyValuePair[k];
      }
    });

    return values;
  }

  inputs = [];
  listeners = [];
  submitting = false;

  getChildContext() {
    return {
      /**
       * Callback function to let the child component register itself on the form.
       * Once registered, the Form component will require the child to have two methods:
       * - handleError
       * - getKeyValue
       *
       * @param {object|Array} props
       * @param {string} props.name
       * @param {*} props.value
       * @param {*} props.values Some components (Slider) uses the plural name
       * @param {object} self The instance.
       */
      register: (props, self) => {
        const { name, value } = props;
        this.inputs = this.inputs || [];
        this.inputs.push({
          defaultValue: value || props.values, // Values is used by slider
          error: null,
          name,
          self
        });
      },

      /**
       * Register a listener for the submit event.
       *
       * @param listener
       */
      onSubmit: listener => {
        this.listeners = this.listeners || [];
        this.listeners.push(listener);
      },

      /**
       * This is the generic error handler. In case the handleSubmit reject
       * passes an _error key, it will be displayed by this component.
       *
       * @param self
       */
      genericErrorHandler: self => {
        this.genericErrorHandler = self;
      }
    };
  }

  handleSubmit = e => {
    if (typeof this.props.handleSubmit !== "function") {
      return false;
    }

    e.preventDefault();

    // Prevent re-submitting form until promise is resolved/rejected.
    if (this.submitting) return false;
    this.submitting = true;

    let values = {};

    // Collect all values from the registered inputs.
    this.inputs.forEach(input => {
      // Reset error state
      if (input.error) {
        input.error = null;
        input.self.handleError(null);
      }

      const { self } = input;
      values = Form.assignKeyPair(values, self.getKeyValue());
    });

    // Make sure to reset the generic error state
    if (this.genericError && this.genericErrorHandler) {
      this.genericErrorHandler.setError(null);
    }

    // Notify listeners and tell them that we are submitting.
    this.listeners.forEach(l => l({ values, loading: true }));

    // Execute the handleSubmit function and wait for a response/reject.
    return this.props
      .handleSubmit(values)
      .catch(errors => {
        if (errors !== null && typeof errors === "object") {
          Object.keys(errors).forEach(key => {
            for (let i = 0; i < this.inputs.length; i++) {
              const input = this.inputs[i];

              if (input.name === key) {
                input.self.handleError(errors[key]);
                input.error = errors[key];
                break;
              }
            }
          });

          if (errors._error && this.genericErrorHandler) {
            this.genericErrorHandler.setError(errors._error);
          }
        }

        // So that we display the shake.
        this.genericError = errors ? !!errors._error : false;
        this.forceUpdate();
      })
      .then(() => {
        this.listeners.forEach(l => l({ values, loading: false }));
        this.submitting = false;
      });
  };

  /**
   * Reset listeners.
   */
  componentWillUnmount() {
    this.listeners = [];
  }

  render() {
    const props = filterProps(Form.propTypes, this.props);
    const formClasses = classNames(this.props.className, {
      [classes.hasError]:
        this.inputs.filter(i => i.error).length > 0 || this.genericError
    });

    return (
      <form {...props} onSubmit={this.handleSubmit} className={formClasses}>
        {this.props.children}
      </form>
    );
  }
}
