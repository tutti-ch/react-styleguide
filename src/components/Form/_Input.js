import React, { Component } from "react";
import PropTypes from "prop-types";
import { filterProps } from "../../helpers/functions";
import WithWrapper from "./_WithWrapper";

// istanbul ignore next
const MutationObserver =
  typeof window !== "undefined"
    ? window.MutationObserver ||
      window.WebKitMutationObserver ||
      window.MozMutationObserver
    : undefined;

export class Input extends Component {
  static defaultProps = {
    value: "",
    type: "text"
  };

  static propTypes = {
    /**
     * The initial value.
     */
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.bool
    ]),

    /**
     * The input type.
     */
    type: PropTypes.string,

    /**
     * The input name.
     */
    name: PropTypes.string
  };

  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };

    this.notify = this.notify.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
  }

  // istanbul ignore next
  componentDidMount() {
    // This code fixes a bug which happens when a user saves information
    // such as login information, and the browser fills the input automatically.
    if (!this.props.value && MutationObserver) {
      this.observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
          const target = mutation.target;

          if (
            target.defaultValue &&
            mutation.type === "attributes" &&
            mutation.attributeName === "defaultValue"
          ) {
            this.changed = true;
            this.setState({ value: target.defaultValue });
            this.notify();
            this.observer = this.observer.disconnect();
            target.defaultValue = ""; // Reset so that react does not complain about controlled inputs
          }
        });
      });

      this.observer.observe(this.refs.input, {
        attributes: true
      });
    }
  }

  // istanbul ignore next
  componentWillUnmount() {
    if (this.observer) {
      this.observer.disconnect();
    }
  }

  /**
   * Sync the value.
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.changed = true;
      this.setState({ value: nextProps.value || "" }, this.notify);
    }
  }

  /**
   * Notify the parent of the change.
   */
  notify() {
    if (this.changed === true && typeof this.props.onChange === "function") {
      const { name, value } = this.props;
      const newValue = this.state.value;
      this.props.onChange(newValue, {
        name,
        initialValue: value,
        formValue: newValue
      });
    }

    // Reset the value
    this.changed = false;
  }

  /**
   * Notify the parent when blur occurs.
   *
   * @param event
   */
  handleBlur(event) {
    this.notify();

    // istanbul ignore else
    if (typeof this.props.onBlur === "function") {
      this.props.onBlur(event);
    }
  }

  /**
   * Notify the parent when user presses enter.
   *
   * @param event
   */
  handleKeyDown(event) {
    if (event.key === "Enter") {
      this.notify();
    }

    if (typeof this.props.onKeyDown === "function") {
      this.props.onKeyDown(event);
    }
  }

  /**
   * Handle the change event.
   */
  handleChange(event) {
    this.setState({ value: event.target.value }, () => {
      // Next time when on blur or key up occurs,
      // they will know that the value changed.
      this.changed = true;
    });
  }

  render() {
    const { value } = this.state;
    const { type, name } = this.props;
    const props = filterProps(Input.propTypes, this.props);

    return (
      <input
        {...props}
        name={name}
        type={type}
        value={value}
        onBlur={this.handleBlur}
        onKeyDown={this.handleKeyDown}
        onChange={this.handleChange}
        ref="input"
      />
    );
  }
}

export default WithWrapper(Input);
