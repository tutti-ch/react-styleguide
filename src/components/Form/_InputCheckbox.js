import React, { Component } from "react"
import PropTypes from "prop-types"
import { filterProps } from "helpers/functions"
import classes from "./Form.scss"
import WithWrapper from "./_WithWrapper"

export class InputCheckbox extends Component {
  static defaultProps = {
    checked: false
  }

  static propTypes = {
    /**
     * Whether the checkbox is checked or not.
     */
    checked: PropTypes.bool,

    /**
     * The value of the input.
     */
    value: PropTypes.any,

    /**
     * The label for the input.
     */
    label: PropTypes.string,
  }

  constructor(props) {
    super(props)

    this.state = {
      checked: props.checked,
    }

    this.toggle = this.toggle.bind(this)
  }

  /**
   * Sync the value.
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.checked !== this.props.checked) {
      this.setState({ checked: nextProps.checked })
    }
  }

  /**
   * Toggle the input checkbox.
   */
  toggle() {
    this.setState({ checked: !this.state.checked }, () => {
      if (typeof this.props.onChange === "function") {
        this.props.onChange(this.state.checked)
      }
    })
  }

  render() {
    const { value, checked } = this.state
    const { label } = this.props
    const props = filterProps(InputCheckbox.propTypes, this.props)

    return (
      // We do not use a real label here because react complains
      // about uncontrolled/controlled components.
      <span onClick={this.toggle} className={classes.cbLabel}>
        <input
          {...props}
          type="checkbox"
          value={value}
          checked={checked}
          onChange={this.toggle}
        />
        <span className={classes.labelText}>
          {label}
        </span>
      </span>
    )
  }
}

export default WithWrapper(InputCheckbox)
