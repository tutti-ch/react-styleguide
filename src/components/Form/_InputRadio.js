import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { filterProps } from "../../helpers/functions";
import classes from "./Form.scss";

export default class InputRadio extends PureComponent {
  static defaultProps = {
    checked: false
  };

  static propTypes = {
    /**
     * Whether the checkbox is checked or not.
     */
    checked: PropTypes.bool,

    /**
     * The value of the input.
     */
    value: PropTypes.any,

    /**
     * The input name.
     */
    name: PropTypes.string,

    /**
     * The label for the input.
     */
    label: PropTypes.string,

    /**
     * The on change callback handler.
     */
    onChange: PropTypes.func
  };

  constructor(props) {
    super(props);

    this.state = {
      checked: props.checked
    };

    this.select = this.select.bind(this);
  }

  /**
   * Sync the value.
   */
  componentDidUpdate(oldProps) {
    if (oldProps.checked !== this.props.checked) {
      this.setState({ checked: this.props.checked });
    }
  }

  /**
   * Select the input radio.
   */
  select() {
    if (this.state.checked) {
      return;
    }

    // Toggling an input radio is not possible
    this.setState({ checked: true }, () => {
      if (typeof this.props.onChange === "function") {
        const { name, checked, value } = this.props;
        this.props.onChange(value, {
          name,
          initialValue: checked,
          formValue: value
        });
      }
    });
  }

  render() {
    const { value, checked } = this.state;
    const { label, name } = this.props;
    const props = filterProps(InputRadio.propTypes, this.props);

    return (
      // We do not use a real label here because react complains
      // about uncontrolled/controlled components.
      <span onClick={this.select} className={classes.cbLabel}>
        <input
          {...props}
          name={name}
          type="radio"
          value={value}
          checked={checked}
          onChange={this.select}
        />
        <span className={classes.labelText}>{label}</span>
      </span>
    );
  }
}
