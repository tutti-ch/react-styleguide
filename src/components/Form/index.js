export { default } from "./_Form";
