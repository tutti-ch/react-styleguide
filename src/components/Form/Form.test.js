/* global describe, expect, test */
import React from "react"
import { mount } from "enzyme"
import Form from "./Form"

describe("(Styleguide) Form", () => {
  test("should match the snapshot", () => {
    const comp = mount(<Form/>)
    expect(comp).toMatchSnapshot()
  })
})