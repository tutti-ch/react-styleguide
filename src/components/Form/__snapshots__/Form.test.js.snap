// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Styleguide) Form should match the snapshot 1`] = `
<Component>
  <Table>
    <table
      className=""
    >
      <Component>
        <tbody>
          <Component>
            <tr>
              <Component>
                <td>
                  <h3>
                    Inputs
                  </h3>
                </td>
              </Component>
              <Component
                colSpan={4}
              >
                <td
                  colSpan={4}
                >
                  <WithWrapper
                    label="Block text input"
                  >
                    <span
                      className=""
                    >
                      <Input
                        onChange={[Function]}
                        type="text"
                        value=""
                      >
                        <input
                          onChange={[Function]}
                          type="text"
                          value=""
                        />
                      </Input>
                      <span
                        className={undefined}
                      >
                        Block text input
                      </span>
                    </span>
                  </WithWrapper>
                  <WithWrapper
                    inline={true}
                    label="Inline text input"
                  >
                    <span
                      className=""
                    >
                      <Input
                        onChange={[Function]}
                        type="text"
                        value=""
                      >
                        <input
                          onChange={[Function]}
                          type="text"
                          value=""
                        />
                      </Input>
                      <span
                        className={undefined}
                      >
                        Inline text input
                      </span>
                    </span>
                  </WithWrapper>
                  <WithWrapper
                    label="Number"
                    max="1000"
                    min="100"
                    step="50"
                    type="number"
                    value="150"
                  >
                    <span
                      className="undefined"
                    >
                      <Input
                        max="1000"
                        min="100"
                        onChange={[Function]}
                        step="50"
                        type="number"
                        value="150"
                      >
                        <input
                          max="1000"
                          min="100"
                          onChange={[Function]}
                          step="50"
                          type="number"
                          value="150"
                        />
                      </Input>
                      <span
                        className={undefined}
                      >
                        Number
                      </span>
                    </span>
                  </WithWrapper>
                  <WithWrapper
                    label="Password"
                    type="password"
                  >
                    <span
                      className=""
                    >
                      <Input
                        onChange={[Function]}
                        type="password"
                        value=""
                      >
                        <input
                          onChange={[Function]}
                          type="password"
                          value=""
                        />
                      </Input>
                      <span
                        className={undefined}
                      >
                        Password
                      </span>
                    </span>
                  </WithWrapper>
                  <WithWrapper
                    label="Email"
                    type="email"
                  >
                    <span
                      className=""
                    >
                      <Input
                        onChange={[Function]}
                        type="email"
                        value=""
                      >
                        <input
                          onChange={[Function]}
                          type="email"
                          value=""
                        />
                      </Input>
                      <span
                        className={undefined}
                      >
                        Email
                      </span>
                    </span>
                  </WithWrapper>
                  <WithWrapper
                    label="Phone"
                    type="phone"
                  >
                    <span
                      className=""
                    >
                      <Input
                        onChange={[Function]}
                        type="phone"
                        value=""
                      >
                        <input
                          onChange={[Function]}
                          type="phone"
                          value=""
                        />
                      </Input>
                      <span
                        className={undefined}
                      >
                        Phone
                      </span>
                    </span>
                  </WithWrapper>
                </td>
              </Component>
            </tr>
          </Component>
          <Component>
            <tr>
              <Component>
                <td>
                  <h3>
                    Slider
                  </h3>
                </td>
              </Component>
              <Component
                colSpan={4}
              >
                <td
                  colSpan={4}
                >
                  <WithWrapper
                    label="Slider"
                    max={10000}
                    maxValue={10000}
                    min={1000}
                    minRange={500}
                    minValue={1300}
                    multiple={true}
                    prefix={
                      Array [
                        "From ",
                        "To ",
                      ]
                    }
                    step={100}
                    suffix=".-"
                  >
                    <span
                      className=""
                    >
                      <Slider
                        crossThumbs={false}
                        label="Slider"
                        max={10000}
                        maxValue={10000}
                        min={1000}
                        minRange={500}
                        minValue={1300}
                        multiple={true}
                        onChange={[Function]}
                        prefix={
                          Array [
                            "From ",
                            "To ",
                          ]
                        }
                        step={100}
                        suffix=".-"
                        template=""
                      >
                        <span
                          className={undefined}
                        >
                          <span
                            className={undefined}
                          >
                            <span
                              className={undefined}
                            >
                              Slider
                            </span>
                            <span
                              className={undefined}
                            >
                              From 1300.-
                            </span>
                            <span
                              className={undefined}
                            >
                              To 10000.-
                            </span>
                          </span>
                          <span
                            className={undefined}
                            name="minValue"
                            onMouseDown={[Function]}
                            onTouchMove={[Function]}
                            onTouchStart={[Function]}
                            style={
                              Object {
                                "left": "3.3333333333333335%",
                              }
                            }
                            tabIndex={1}
                          />
                          <span
                            className={undefined}
                            name="maxValue"
                            onMouseDown={[Function]}
                            onTouchMove={[Function]}
                            onTouchStart={[Function]}
                            style={
                              Object {
                                "left": "100%",
                              }
                            }
                            tabIndex={1}
                          />
                        </span>
                      </Slider>
                    </span>
                  </WithWrapper>
                </td>
              </Component>
            </tr>
          </Component>
          <Component>
            <tr>
              <Component>
                <td>
                  <h3>
                    Checkbox Group
                  </h3>
                </td>
              </Component>
              <Component
                colSpan={4}
              >
                <td
                  colSpan={4}
                >
                  <WithWrapper
                    checked={true}
                    label="Red"
                    value="red"
                  >
                    <span
                      className="undefined"
                    >
                      <InputCheckbox
                        checked={true}
                        label="Red"
                        onChange={[Function]}
                        value="red"
                      >
                        <span
                          className={undefined}
                          onClick={[Function]}
                        >
                          <input
                            checked={true}
                            onChange={[Function]}
                            type="checkbox"
                            value={undefined}
                          />
                          <span
                            className={undefined}
                          >
                            Red
                          </span>
                        </span>
                      </InputCheckbox>
                    </span>
                  </WithWrapper>
                  <WithWrapper
                    checked={true}
                    label="Blue"
                    value="blue"
                  >
                    <span
                      className="undefined"
                    >
                      <InputCheckbox
                        checked={true}
                        label="Blue"
                        onChange={[Function]}
                        value="blue"
                      >
                        <span
                          className={undefined}
                          onClick={[Function]}
                        >
                          <input
                            checked={true}
                            onChange={[Function]}
                            type="checkbox"
                            value={undefined}
                          />
                          <span
                            className={undefined}
                          >
                            Blue
                          </span>
                        </span>
                      </InputCheckbox>
                    </span>
                  </WithWrapper>
                  <WithWrapper
                    label="Green"
                    value="green"
                  >
                    <span
                      className="undefined"
                    >
                      <InputCheckbox
                        checked={false}
                        label="Green"
                        onChange={[Function]}
                        value="green"
                      >
                        <span
                          className={undefined}
                          onClick={[Function]}
                        >
                          <input
                            checked={false}
                            onChange={[Function]}
                            type="checkbox"
                            value={undefined}
                          />
                          <span
                            className={undefined}
                          >
                            Green
                          </span>
                        </span>
                      </InputCheckbox>
                    </span>
                  </WithWrapper>
                </td>
              </Component>
            </tr>
          </Component>
          <Component>
            <tr>
              <Component>
                <td>
                  <h3>
                    Button Group
                  </h3>
                </td>
              </Component>
              <Component
                colSpan={4}
              >
                <td
                  colSpan={4}
                >
                  <InputGroup>
                    <WithWrapper
                      checked={true}
                      key=".0"
                      label="React.js"
                      name="r1"
                      onChange={[Function]}
                      value={10}
                    >
                      <span
                        className="undefined"
                      >
                        <InputRadio
                          checked={true}
                          label="React.js"
                          name="r1"
                          onChange={[Function]}
                          value={10}
                        >
                          <span
                            className={undefined}
                            onClick={[Function]}
                          >
                            <input
                              checked={true}
                              name="r1"
                              onChange={[Function]}
                              type="radio"
                              value={undefined}
                            />
                            <span
                              className={undefined}
                            >
                              React.js
                            </span>
                          </span>
                        </InputRadio>
                      </span>
                    </WithWrapper>
                    <WithWrapper
                      checked={false}
                      key=".1"
                      label="Vue.js"
                      name="r1"
                      onChange={[Function]}
                      value={10}
                    >
                      <span
                        className="undefined"
                      >
                        <InputRadio
                          checked={false}
                          label="Vue.js"
                          name="r1"
                          onChange={[Function]}
                          value={10}
                        >
                          <span
                            className={undefined}
                            onClick={[Function]}
                          >
                            <input
                              checked={false}
                              name="r1"
                              onChange={[Function]}
                              type="radio"
                              value={undefined}
                            />
                            <span
                              className={undefined}
                            >
                              Vue.js
                            </span>
                          </span>
                        </InputRadio>
                      </span>
                    </WithWrapper>
                    <WithWrapper
                      checked={false}
                      key=".2"
                      label="Angular.js"
                      name="r1"
                      onChange={[Function]}
                      value={10}
                    >
                      <span
                        className="undefined"
                      >
                        <InputRadio
                          checked={false}
                          label="Angular.js"
                          name="r1"
                          onChange={[Function]}
                          value={10}
                        >
                          <span
                            className={undefined}
                            onClick={[Function]}
                          >
                            <input
                              checked={false}
                              name="r1"
                              onChange={[Function]}
                              type="radio"
                              value={undefined}
                            />
                            <span
                              className={undefined}
                            >
                              Angular.js
                            </span>
                          </span>
                        </InputRadio>
                      </span>
                    </WithWrapper>
                  </InputGroup>
                </td>
              </Component>
            </tr>
          </Component>
          <Component>
            <tr>
              <Component>
                <td>
                  <h3>
                    Textarea
                  </h3>
                </td>
              </Component>
              <Component
                colSpan={4}
              >
                <td
                  colSpan={4}
                >
                  <WithWrapper
                    label="This is a textarea"
                  >
                    <span
                      className=""
                    >
                      <Textarea
                        onChange={[Function]}
                        value=""
                      >
                        <textarea
                          onChange={[Function]}
                          value=""
                        />
                      </Textarea>
                      <span
                        className={undefined}
                      >
                        This is a textarea
                      </span>
                    </span>
                  </WithWrapper>
                </td>
              </Component>
            </tr>
          </Component>
        </tbody>
      </Component>
    </table>
  </Table>
</Component>
`;
